### DISCUSSION TITLE
ScriptReference/UIElements.CallbackEventHandler.RegisterCallback

### DISCUSSION THREAD

**Discussion by giscus** [Developer-Notes-Extension/Unity-Notes#548] [Notes]
# ScriptReference/UIElements.CallbackEventHandler.RegisterCallback

https://docs.unity3d.com/ScriptReference/UIElements.CallbackEventHandler.RegisterCallback.html

**Comment by vertxxyz** [comment]
You can see a curated list of events [here](https://docs.unity3d.com/Manual/UIE-Events-Reference.html).
I have compiled the list for reference purposes:

### Events in `UnityEngine` and `UnityEditor` as of 2023.2:
#### UnityEngine.UIElements
| Type  | Description |
| ------------- | ------------- |
| [`AttachToPanelEvent`](https://docs.unity3d.com/ScriptReference/UIElements.AttachToPanelEvent.html) | When an element is added to an element that is a descendent of a panel. |
| [`BlurEvent`](https://docs.unity3d.com/ScriptReference/UIElements.BlurEvent.html) | Sent immediately after an element has lost focus. This event trickles down, it does not bubble up, and it cannot be cancelled. |
| [`ChangeEvent<T0>`](https://docs.unity3d.com/ScriptReference/UIElements.ChangeEvent_1.html) | When a value in a field changes. |
| [`ClickEvent`](https://docs.unity3d.com/ScriptReference/UIElements.ClickEvent.html) | When the left mouse button is clicked. |
| [`ContextClickEvent`](https://docs.unity3d.com/ScriptReference/UIElements.ContextClickEvent.html) | When clicking the right mouse button. |
| [`ContextualMenuPopulateEvent`](https://docs.unity3d.com/ScriptReference/UIElements.ContextualMenuPopulateEvent.html) | When a contextual menu requires menu items. |
| [`CustomStyleResolvedEvent`](https://docs.unity3d.com/ScriptReference/UIElements.CustomStyleResolvedEvent.html) | When the custom style properties of a VisualElement have been resolved. |
| [`DetachFromPanelEvent`](https://docs.unity3d.com/ScriptReference/UIElements.DetachFromPanelEvent.html) | Sent just before an element is detached from its parent if the parent is the descendant of a panel. |
| [`DragEnterEvent`](https://docs.unity3d.com/ScriptReference/UIElements.DragEnterEvent.html) | Used to manage events that occur when dragging enters an element or one of its descendants. The DragEnterEvent does not trickle down and does not bubble up. |
| [`DragExitedEvent`](https://docs.unity3d.com/ScriptReference/UIElements.DragExitedEvent.html) | Sent to a dragged element when the drag and drop process ends. |
| [`DragLeaveEvent`](https://docs.unity3d.com/ScriptReference/UIElements.DragLeaveEvent.html) | Used to manage events sent when dragging leaves an element or one of its descendants. The DragLeaveEvent does not trickle down and does not bubble up. |
| [`DragPerformEvent`](https://docs.unity3d.com/ScriptReference/UIElements.DragPerformEvent.html) | When another element is dragged and dropped on the element. |
| [`DragUpdatedEvent`](https://docs.unity3d.com/ScriptReference/UIElements.DragUpdatedEvent.html) | When the element being dragged enters a possible drop target. |
| [`ExecuteCommandEvent`](https://docs.unity3d.com/ScriptReference/UIElements.ExecuteCommandEvent.html) | Sent by the Editor when an element in the panel should execute a command. (See also [command events](https://docs.unity3d.com/Manual/UIE-Command-Events.html)) |
| [`FocusEvent`](https://docs.unity3d.com/ScriptReference/UIElements.FocusEvent.html) | Immediately after an element has gained focus. This event trickles down and does not bubbles up. |
| [`FocusInEvent`](https://docs.unity3d.com/ScriptReference/UIElements.FocusInEvent.html) | Immediately before an element gains focus. This event trickles down and bubbles up. |
| [`FocusOutEvent`](https://docs.unity3d.com/ScriptReference/UIElements.FocusOutEvent.html) | Immediately before an element loses focus. This event trickles down and bubbles up. |
| [`GeometryChangedEvent`](https://docs.unity3d.com/ScriptReference/UIElements.GeometryChangedEvent.html) | Sent after layout calculations, when the position or the dimension of an element changes. |
| [`IMGUIEvent`](https://docs.unity3d.com/ScriptReference/UIElements.IMGUIEvent.html) | Used to send a IMGUI event that has no equivalent UIElements event. |
| [`InputEvent`](https://docs.unity3d.com/ScriptReference/UIElements.InputEvent.html) | When text from a TextField changes. |
| [`KeyDownEvent`](https://docs.unity3d.com/ScriptReference/UIElements.KeyDownEvent.html) | When a key is pressed. |
| [`KeyUpEvent`](https://docs.unity3d.com/ScriptReference/UIElements.KeyUpEvent.html) | When a pressed key is released. |
| [`MouseCaptureEvent`](https://docs.unity3d.com/ScriptReference/UIElements.MouseCaptureEvent.html) | After a handler starts capturing the mouse. |
| [`MouseCaptureOutEvent`](https://docs.unity3d.com/ScriptReference/UIElements.MouseCaptureOutEvent.html) | Before a handler stops capturing the mouse. |
| [`MouseDownEvent`](https://docs.unity3d.com/ScriptReference/UIElements.MouseDownEvent.html) | When a mouse button is pressed. |
| [`MouseEnterEvent`](https://docs.unity3d.com/ScriptReference/UIElements.MouseEnterEvent.html) | When the mouse pointer enters an element or one of its descendent elements. The event trickles down but does not bubble up. |
| [`MouseEnterWindowEvent`](https://docs.unity3d.com/ScriptReference/UIElements.MouseEnterWindowEvent.html) | When the mouse pointer enters a window. The event bubbles up but does not trickle down. |
| [`MouseLeaveEvent`](https://docs.unity3d.com/ScriptReference/UIElements.MouseLeaveEvent.html) | When the mouse pointer exits an element and all its descendent elements. The event trickles down but does not bubble up. |
| [`MouseLeaveWindowEvent`](https://docs.unity3d.com/ScriptReference/UIElements.MouseLeaveWindowEvent.html) | When the mouse pointer exits a window. The event bubbles up but does not trickle down. |
| [`MouseMoveEvent`](https://docs.unity3d.com/ScriptReference/UIElements.MouseMoveEvent.html) | When the mouse moves. |
| [`MouseOutEvent`](https://docs.unity3d.com/ScriptReference/UIElements.MouseOutEvent.html) | When the mouse pointer exits an element. The event trickles down and bubbles up. |
| [`MouseOverEvent`](https://docs.unity3d.com/ScriptReference/UIElements.MouseOverEvent.html) | When the mouse pointer enters an element. The event trickles down and bubbles up. |
| [`MouseUpEvent`](https://docs.unity3d.com/ScriptReference/UIElements.MouseUpEvent.html) | When a mouse button is released. |
| [`NavigationCancelEvent`](https://docs.unity3d.com/ScriptReference/UIElements.NavigationCancelEvent.html) | When the user presses the cancel button. |
| [`NavigationMoveEvent`](https://docs.unity3d.com/ScriptReference/UIElements.NavigationMoveEvent.html) | Typically when the user presses the D-pad, moves a joystick or presses the arrow keys. |
| [`NavigationSubmitEvent`](https://docs.unity3d.com/ScriptReference/UIElements.NavigationSubmitEvent.html) | When the user presses the submit button. |
| [`PointerCancelEvent`](https://docs.unity3d.com/ScriptReference/UIElements.PointerCancelEvent.html) | When pointer interaction is cancelled. |
| [`PointerCaptureEvent`](https://docs.unity3d.com/ScriptReference/UIElements.PointerCaptureEvent.html) | When a pointer is captured by a VisualElement. |
| [`PointerCaptureOutEvent`](https://docs.unity3d.com/ScriptReference/UIElements.PointerCaptureOutEvent.html) | When a VisualElement releases a pointer. |
| [`PointerDownEvent`](https://docs.unity3d.com/ScriptReference/UIElements.PointerDownEvent.html) | When a pointer is pressed. |
| [`PointerEnterEvent`](https://docs.unity3d.com/ScriptReference/UIElements.PointerEnterEvent.html) | When a pointer enters a VisualElement or one of its descendants. The event does not trickle down and does not bubble up. |
| [`PointerLeaveEvent`](https://docs.unity3d.com/ScriptReference/UIElements.PointerLeaveEvent.html) | Sent when a pointer exits an element and all of its descendants. The event does not trickle down and does not bubble up. |
| [`PointerMoveEvent`](https://docs.unity3d.com/ScriptReference/UIElements.PointerMoveEvent.html) | When a pointer changes state. For example if its position or pressure change, or a different button is pressed. |
| [`PointerOutEvent`](https://docs.unity3d.com/ScriptReference/UIElements.PointerOutEvent.html) | When a pointer exits an element. The event trickles down and bubbles up. |
| [`PointerOverEvent`](https://docs.unity3d.com/ScriptReference/UIElements.PointerOverEvent.html) | When a pointer enters an element. The event trickles down and bubbles up. |
| [`PointerStationaryEvent`](https://docs.unity3d.com/ScriptReference/UIElements.PointerStationaryEvent.html) | When a pointer does not change for a set amount of time, determined by the operating system. |
| [`PointerUpEvent`](https://docs.unity3d.com/ScriptReference/UIElements.PointerUpEvent.html) | When a pointer's last pressed button is released. |
| [`TooltipEvent`](https://docs.unity3d.com/ScriptReference/UIElements.TooltipEvent.html) | Sent to find the first VisualElement that displays a tooltip. |
| [`TransitionCancelEvent`](https://docs.unity3d.com/ScriptReference/UIElements.TransitionCancelEvent.html) | When a transition is canceled. |
| [`TransitionEndEvent`](https://docs.unity3d.com/ScriptReference/UIElements.TransitionEndEvent.html) | The completion of a transition. In the case where a transition is removed before completion then the event will not fire. |
| [`TransitionRunEvent`](https://docs.unity3d.com/ScriptReference/UIElements.TransitionRunEvent.html) | When a transition is created (i.e. added to the set of running transitions). |
| [`TransitionStartEvent`](https://docs.unity3d.com/ScriptReference/UIElements.TransitionStartEvent.html) | When a transition's delay phase ends. |
| [`ValidateCommandEvent`](https://docs.unity3d.com/ScriptReference/UIElements.ValidateCommandEvent.html) | Sent by the Editor while it determines whether the command will be handled by an element in the panel. |
| [`WheelEvent`](https://docs.unity3d.com/ScriptReference/UIElements.WheelEvent.html) | When the mouse wheel moves. |

#### UnityEditor.UIElements
| Type  | Description |
| ------------- | ------------- |
| [`SerializedObjectChangeEvent`](https://docs.unity3d.com/ScriptReference/UIElements.SerializedObjectChangeEvent.html) | When any value in a SerializedObject changes. |
| [`SerializedPropertyChangeEvent`](https://docs.unity3d.com/ScriptReference/UIElements.SerializedPropertyChangeEvent.html) | When a value in a PropertyField changes. |
| `SerializedObjectBindEvent` **(internal)** | When a SerializedObject is bound to the element. |
| `SerializedPropertyBindEvent` **(internal)** | When a SerializedProperty is bound to the element. |

**Comment by vertxxyz** [comment]
There exists a rather large trap when using the callback that takes `userArgs` that means only one event is registered on a target despite registering two:
> [**Handle events**](https://docs.unity3d.com/Manual/UIE-Events-Handling.html)
> You can register multiple callbacks for an event. However, you can only register the same callback function on the same event and propagation phase once.

"The same callback function" is comparing delegates, and multiple calls to an event like `target.RegisterCallback<Event, This>((evt, args) => args.Callback(evt), this);` will be the same delegate, and subsequent calls will just overwrite the old ones.
However multiple calls to `target.RegisterCallback<Event>(evt => Callback(evt));` involves different delegates, as an argument is captured.
So, use of the `userArgs`-taking method like this (one that isn't still involving a capture) will only allow a single event of the same type to register on one target.
